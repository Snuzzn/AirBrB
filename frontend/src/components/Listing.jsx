import React from 'react';
import PropTypes from 'prop-types';
import { MdOutlineRateReview } from 'react-icons/md';

const Listing = ({ thumbnail, reviews = [], title, price }) => {
  const [showTooltip, setShowTooltip] = React.useState(false);

  return (
    <>
      <div className="relative -mb-8 text-right pr-4 cursor">
        <span id="listing-price" className="bg-red-400 cursor-default text-white rounded-xl transit pb-1 pt-1 pl-2 pr-2 transition duration-500 ease-in-out hover:text-red-400 hover:bg-white">${price}</span>
      </div>
      <div className="">
        <img className="object-cover h-60 w-full rounded-2xl" src={thumbnail} alt="Image of a your listing" />
      </div>
      <div className="relative mt-1 flex justify-between pl-2 pr-2">
        <div className="text-lg">
          {title}
        </div>
        <div
          className="flex flex-row text-center items-center text-gray-500"
          onMouseEnter={() => setShowTooltip(true)}
          onMouseLeave={() => setShowTooltip(false)}
        >
          <span
            className="inline-block pr-1 text-lg"
            aria-label="Number of reviews"
            id="review-length"
          >
            {reviews.length}
          </span>
          <MdOutlineRateReview id="review-number-icon" aria-label="Number of reviews" title="Icon representing number of reviews" />
          {showTooltip && <div className="absolute rounded text-xs shadow-lg p-1 bg-gray-100 text-black bottom-7">Review Count</div>}
        </div>
      </div>
    </>
  )
}

export default Listing;

Listing.propTypes = {
  thumbnail: PropTypes.string,
  reviews: PropTypes.array,
  title: PropTypes.string,
  price: PropTypes.number
}
