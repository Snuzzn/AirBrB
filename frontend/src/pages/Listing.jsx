import React from 'react';
import Fade from 'react-reveal/Fade';

const Listings = () => {
  return (
    <Fade>
      <div>Hello</div>
    </Fade>
  )
}

export default Listings;
