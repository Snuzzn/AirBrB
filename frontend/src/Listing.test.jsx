import React from 'react';
import Listing from './components/Listing';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

describe('Listing', () => {
  /**
   * Listing takes in a thumbnail, reviews,
   * title and price as props, and displays all but
   * the id prop.
   */

  it('uses thumbnail string', () => {
    const listing = shallow(<Listing
      thumbnail="test image"
    />);
    expect(listing.find('img').prop('src')).toBe('test image');
  })

  it('thumbnail alt prop is defined for accessibility', () => {
    const listing = shallow(<Listing
      thumbnail="test image"
    />);
    expect(listing.find('img').prop('alt')).toBeDefined();
  })

  it('rendered with length of review list', () => {
    const reviews = [
      {
        score: 5,
        review: 'Great place!'
      },
      {
        score: 1,
        review: 'Not enough almond milk'
      }
    ]

    const listing = shallow(<Listing
      reviews={reviews}
    />);
    expect(listing.find('#review-length').text()).toBe('2');
  })

  it('uses default value of 0 if no reviews', () => {
    const listing = shallow(<Listing/>);
    expect(listing.find('#review-length').text()).toBe('0');
  })

  it('rendered with correctly formatted price of listing', () => {
    const price = 150;
    const listing = shallow(<Listing
      price={price}
    />);
    // Price should have a '$' preceding it
    expect(listing.find('#listing-price').text()).toBe('$150');
  })

  it('review number icon has title for accessibility', () => {
    const listing = shallow(<Listing />);
    expect(listing.find('#review-number-icon').prop('title')).toBeDefined();
  })

  it('review number icon has aria-label for accessibility', () => {
    const listing = shallow(<Listing />);
    expect(listing.find('#review-number-icon').prop('aria-label')).toBeDefined();
  })

  // Snapshots
  it('renders with minimal props', () => {
    const listing = renderer.create(<Listing />).toJSON();
    expect(listing).toMatchSnapshot();
  })

  it('renders with provided props', () => {
    const reviews = [
      {
        score: 5,
        review: 'Great place!'
      },
      {
        score: 1,
        review: 'Not enough almond milk'
      }
    ]

    const listing = renderer.create(<Listing
      thumbnail='I am a thumbnail!'
      reviews={reviews}
      title={'Cool house!'}
      price={150}
    />).toJSON();
    expect(listing).toMatchSnapshot();
  })
})
