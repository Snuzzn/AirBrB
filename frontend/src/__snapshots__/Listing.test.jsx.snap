// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Listing renders with minimal props 1`] = `
Array [
  <div
    className="relative -mb-8 text-right pr-4 cursor"
  >
    <span
      className="bg-red-400 cursor-default text-white rounded-xl transit pb-1 pt-1 pl-2 pr-2 transition duration-500 ease-in-out hover:text-red-400 hover:bg-white"
      id="listing-price"
    >
      $
    </span>
  </div>,
  <div
    className=""
  >
    <img
      alt="Image of a your listing"
      className="object-cover h-60 w-80 rounded-2xl"
    />
  </div>,
  <div
    className="relative mt-1 flex justify-between pl-2 pr-2"
  >
    <div
      className="text-lg"
    />
    <div
      className="flex flex-row text-center items-center text-gray-500"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <span
        aria-label="Number of reviews"
        className="inline-block pr-1 text-lg"
        id="review-length"
      >
        0
      </span>
      <svg
        fill="currentColor"
        height="1em"
        id="review-number-icon"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          Icon representing number of reviews
        </title>
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M20 2H4c-1.1 0-1.99.9-1.99 2L2 22l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm0 14H5.17l-.59.59-.58.58V4h16v12zm-9.5-2H18v-2h-5.5zm3.86-5.87c.2-.2.2-.51 0-.71l-1.77-1.77c-.2-.2-.51-.2-.71 0L6 11.53V14h2.47l5.89-5.87z"
        />
      </svg>
    </div>
  </div>,
]
`;

exports[`Listing renders with provided props 1`] = `
Array [
  <div
    className="relative -mb-8 text-right pr-4 cursor"
  >
    <span
      className="bg-red-400 cursor-default text-white rounded-xl transit pb-1 pt-1 pl-2 pr-2 transition duration-500 ease-in-out hover:text-red-400 hover:bg-white"
      id="listing-price"
    >
      $
      150
    </span>
  </div>,
  <div
    className=""
  >
    <img
      alt="Image of a your listing"
      className="object-cover h-60 w-80 rounded-2xl"
      src="I am a thumbnail!"
    />
  </div>,
  <div
    className="relative mt-1 flex justify-between pl-2 pr-2"
  >
    <div
      className="text-lg"
    >
      Cool house!
    </div>
    <div
      className="flex flex-row text-center items-center text-gray-500"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <span
        aria-label="Number of reviews"
        className="inline-block pr-1 text-lg"
        id="review-length"
      >
        2
      </span>
      <svg
        fill="currentColor"
        height="1em"
        id="review-number-icon"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          Icon representing number of reviews
        </title>
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M20 2H4c-1.1 0-1.99.9-1.99 2L2 22l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm0 14H5.17l-.59.59-.58.58V4h16v12zm-9.5-2H18v-2h-5.5zm3.86-5.87c.2-.2.2-.51 0-.71l-1.77-1.77c-.2-.2-.51-.2-.71 0L6 11.53V14h2.47l5.89-5.87z"
        />
      </svg>
    </div>
  </div>,
]
`;
